Fix out of bound reads and line numbers in CaosLexer for byte strings

mtlynch opened the pull request:
This includes three changes:

1. Add unit tests for how the lexer currently handles byte string tokens
2. Fix a bug in the lexer that caused it to read beyond the input buffer's allocated memory
3. Adjust the lexer's behavior for seeing newlines in byte strings so that newlines cause it to increment the line number associated with subsequent tokens, even though the newline within the byte strings are an error

Note that the lexer will treat a Windows newline (`\r\n`) as two lines, but that's consistent with how it treats regular strings. We can fix that in a subsequent PR, but I want to avoid expanding the scope of this one.

mtlynch committed:
Fix out of bound read in CaosLexer - unterminated byte string

This fixes a bug in the CAOS lexer where it accidentally reads beyond the null termination character for a malformed CAOS input.

Specifically, if CAOS has an unterminated byte string like '[', but the input ends before any characters, the lexer gets confused and reads beyond the buffer.

This fixes the lexer bug and adds a unit test to exercise the fix.

ligfx reviewed (changes requested):
Looks good to me overall:

- I've changed the formatting of error tokens so that they actually output the string they cover, which will break these tests. Can you rebase and update this?

- This makes me realize that the line numbers should be incremented when running into erroneous newlines/carriage returns. Would you mind fixing that?

- ~~I'm also curious what the behavior of `[01\r\n02]` is. It looks like it will return `TOK_ERROR([01\r), TOK_NEWLINE(\n)` but it should probably return `TOK_ERROR([01\r\n)` (and have the correct line numbers to match that). Would you mind taking a crack at that?~~

Feel free to combine this with PR #221 as well if it's easier to do it all in one place.

ligfx commented:
> I'm also curious what the behavior of [01\r\n02] is. It looks like it will return TOK_ERROR([01\r), TOK_NEWLINE(\n) but it should probably return TOK_ERROR([01\r\n) (and have the correct line numbers to match that). Would you mind taking a crack at that?

Actually, never mind on this. I don't think it's important enough to matter, and errors should be reported correctly either way.

mtlynch committed:
Add unit tests for byte strings in CAOS lexer

mtlynch committed:
In CAOS lexer, increment line number for newlines within byte string

Previously, the if the lexer encountered a newline or carriage return within a bytestring, it wouldn't increment the line count of the tokens even though they technically appeared on a later line of input. This updates the line counter so that newlines or carriage returns within a byte string increment the line count of the lexer's tokens.

mtlynch commented:
>I've changed the formatting of error tokens so that they actually output the string they cover, which will break these tests. Can you rebase and update this?

Sure, I've rebased and updated the tests. I think that improves the output.

>This makes me realize that the line numbers should be incremented when running into erroneous newlines/carriage returns. Would you mind fixing that?

Sure, I've fixed that.

I merged this PR with #221 because incrementing the line numbers made more sense if I did it on top of 4b73f45b0059cdf2b860798dc6fb4a1b4274ce10, but I've kept the commits separate for a clean history.

ligfx commented on src/fileformats/caoslexer.cpp:
```diff
@@ -205,6 +205,8 @@ void lexcaos(std::vector<caostoken>& v, const char* p) {
 		push_value(caostoken::TOK_ERROR);
 	} else if (p[0] == '\r' || p[0] == '\n') {
 		p++;
+		yylineno++;
```
Only `\n` (or `\r\n` if treated as an atomic unit) is a newline and should increment the line number. As is, this would cause an erroneous double line increment for `[\r\n`

ligfx reviewed (changes requested):
- `\r` shouldn’t increment the line number, it’s just whitespace (unless treated as an atomic unit with `\r\n`) (there _are_ some old Mac OS files out there from Norngarden that use `\r` as a newline, but they’re also in a completely different text encoding, MacRoman, that isn’t even handled by the official engine, so meh)
- Can you fix the line numbers before adding the tests? I’d rather not have tests in the commit history that test that we get incorrect values.
- Could you fix the line numbers for quoted strings too while we’re here?

Thanks!

mtlynch commented:
>Can you fix the line numbers before adding the tests? I’d rather not have tests in the commit history that test that we get incorrect values.

Generally, when working with code that doesn't have code coverage, I think the best order of operations is:

1. Add tests while holding the production code constant
2. (optional) Refactor the production code while keeping the test code constant
3. Make behavior changes and update the tests to match

Doing the changes separately makes the change history easy to understand because you can have confidence that behavior didn't change in steps (1) or (2). And then in the change for step (3), you can see exactly how the behavior change impacted the tests.

I understand why you wouldn't want the HEAD commit in main to have tests asserting undesirable behavior, but we're avoiding that here because they'll all be merged at once so HEAD will always have the bugfix.

The incorrect line numbers were the lexer's actual behavior before df413f86cd5dd3d1045dc90a8d809a82f85a04db, so why not make that behavior clear in the source history? That way, df413f86cd5dd3d1045dc90a8d809a82f85a04db shows the fix and shows the effect on the tests. If I go back and make the fix before adding the baseline tests, it's hard to see what the bugfix changed.

Is there a situation you have in mind where it would cause problems to have unit tests in the source history demonstrating the previous line count behavior?

ligfx commented:
I disagree. Tests should check for correct behavior, not current behavior.

mtlynch commented:
Okay, I think I'm going to disengage then.

I appreciate you reviewing my patches, but I feel like you're doing a lot of things to make it unnecessarily difficult to contribute, and it sounds like we're pretty far apart on development philosophy.

ligfx commented:
Sorry to hear it, @mtlynch. Thanks for your contributions thus far.
